src/graph/sampler.cc

```diff
@@ -524,4 +525,198 @@ IdArray SamplerOp::RandomWalk(
   return traces;
 }
 
+void ConstructLayers(const int64_t *indptr,
+                     const dgl_id_t *indices,
+                     const IdArray seed_array,
+                     const std::vector<size_t> &layer_sizes,
+                     std::vector<dgl_id_t> *layer_offsets,
+                     std::vector<dgl_id_t> *node_mapping,
+                     std::vector<int64_t> *actl_layer_sizes,
+                     std::vector<float> *probabilities) {
+  const dgl_id_t* seed_data = static_cast<dgl_id_t*>(seed_array->data);
+  size_t seed_len = seed_array->shape[0];
+  std::copy(seed_data, seed_data + seed_len, std::back_inserter(*node_mapping));
+  actl_layer_sizes->push_back(node_mapping->size());
+  probabilities->insert(probabilities->end(), node_mapping->size(), 1);
+
+  size_t curr = 0;
+  size_t next = node_mapping->size();
+  unsigned int rand_seed = time(nullptr);
+  for (auto i = layer_sizes.rbegin(); i != layer_sizes.rend(); ++i) {
+    std::unordered_set<dgl_id_t> candidate_set;
+    for (auto j = curr; j != next; ++j) {
+      auto src = (*node_mapping)[j];
+      candidate_set.insert(indices + indptr[src], indices + indptr[src + 1]);
+    }
+
+    std::vector<dgl_id_t> candidate_vector;
+    std::copy(candidate_set.begin(), candidate_set.end(),
+              std::back_inserter(candidate_vector));
+
+    std::unordered_map<dgl_id_t, size_t> n_occurrences;
+    auto n_candidates = candidate_vector.size();
+    for (size_t j = 0; j != *i; ++j) {
+      auto dst = candidate_vector[rand_r(&rand_seed) % n_candidates];
+      if (!n_occurrences.insert(std::make_pair(dst, 1)).second) {
+        ++n_occurrences[dst];
+      }
```

Comment:
so it doesn't sample the specified number of vertices in a layer?

Reply:
The sampling is with replacement, which is required by the Monte Carlo method. So the number of sampled nodes may be less than `layer_size`. The number of times that a node is sampled is reflected in its corresponding sampling probability.

Reply:
it should work for both with or without replacement. @ZiyueHuang could you confirm?

Reply:
neighbor sampling samples neighbors without replacement.